File: dotnet/RailsLib.Net/Net/Algorithms/NetworkIterator.cs
using GameLib.Net.Common;
using GameLib.Net.Game;
using GameLib.Net.Graph;
using System;
using System.Collections.Generic;
using System.Text;

namespace GameLib.Net.Algorithms
{
    public class NetworkIterator : GraphEnumerator<NetworkVertex, NetworkEdge>//, IEnumerator<NetworkVertex>
    {
        public enum GreedyState
        {
            seen,
            nonGreedy,
            greedy,
            done
        }

        // settings
        private NetworkVertex startVertex;
        private bool startVertexVisited;
        private bool routeIterator;

        // internal data
        private List<NetworkVertex> stack = new List<NetworkVertex>();
        private List<bool> greedyStack = new List<bool>();
        private Dictionary<NetworkVertex, GreedyState> seen = new Dictionary<NetworkVertex, GreedyState>();


        private IGraph<NetworkVertex, NetworkEdge> graph;

        protected static Logger<NetworkIterator> log = new Logger<NetworkIterator>();

    
    public NetworkIterator(IGraph<NetworkVertex, NetworkEdge> graph,
            NetworkVertex startVertex) : this(graph, startVertex, null)
        {
            
        }

        /**
         * Returns NetworkIterator for specific company
         */
        public NetworkIterator(IGraph<NetworkVertex, NetworkEdge> graph, NetworkVertex startVertex, PublicCompany company) : base()
        {
            if (graph == null)
                throw new ArgumentException("graph must not be null");

            if (!graph.ContainsVertex(startVertex))
                throw new ArgumentException("graph must contain the start vertex");

            this.graph = graph;
            this.startVertex = startVertex;
            this.startVertexVisited = false;
            this.routeIterator = false;
        }

        public NetworkIterator SetRouteIterator(bool routeIterator)
        {
            this.routeIterator = routeIterator;
            return this;
        }

        /**
         * @return the graph being traversed
         */
        public IGraph<NetworkVertex, NetworkEdge> Graph
        {
            get
            {
                return graph;
            }
        }

        public Dictionary<NetworkVertex, GreedyState> SeenData
        {
            get
            {
                return seen;
            }
        }

        public NetworkVertex Current { get; set; }

        public List<NetworkVertex> GetCurrentRoute()
        {
            // extract all networkvertices just before a null
            List<NetworkVertex> route = new List<NetworkVertex>();
            NetworkVertex previousVertex = null;
            foreach (NetworkVertex vertex in stack)
            {
                if (previousVertex != null && vertex == null)
                {
                    route.Add(previousVertex);
                }
                previousVertex = vertex;
            }
            return route;
        }

        /**
         * @see java.util.Iterator#hasNext()
         */
        private bool HasNext()
        {
            if (!startVertexVisited)
            {
                EncounterStartVertex();
            }

            int i = stack.Count - 1;
            while (i >= 0)
            {
                if (stack[i] != null)
                    break;
                else
                    i = i - 2;
            }
            return i >= 0;
        }

        /**
         * @see java.util.Iterator#next()
         */
        public /*NetworkVertex*/bool MoveNext()
        {
            if (!startVertexVisited)
            {
                EncounterStartVertex();
            }

            if (HasNext())
            {
                NetworkVertex nextVertex;
                while (true)
                {
                    nextVertex = stack[stack.Count - 1];
                    stack.RemoveAt(stack.Count - 1);

                    if (nextVertex != null)
                        break;
                    stack.RemoveAt(stack.Count - 1);
                }

                log.Debug("Iterator: provides next vertex" + nextVertex);
                bool nextGreedy = greedyStack[greedyStack.Count - 1];
                    greedyStack.RemoveAt(greedyStack.Count - 1);

                PutSeenData(nextVertex, nextGreedy);
                stack.Add(nextVertex);
                stack.Add(null); // add sentinel that we know when we are ready

                AddUnseenChildrenOf(nextVertex, nextGreedy);

                //return nextVertex;
                Current = nextVertex;
                return true;
            }
            else
            {
                //throw new NoSuchElementException();
                Current = null;
                return false;
            }
        }

        private void PutSeenData(NetworkVertex vertex, bool greedy)
        {
            if (!vertex.IsSide)
            {
                seen[vertex] = GreedyState.seen;
                log.Debug("Iterator:  Vertex " + vertex + " seen with greedyState = seen");
                return;
            }
            // side
            if (seen.ContainsKey(vertex))
            {
                seen[vertex] = GreedyState.done;
                log.Debug("Iterator:  Vertex " + vertex + " seen with greedyState = done");
            }
            else if (greedy)
            {
                seen[vertex] = GreedyState.greedy;
                log.Debug("Iterator:  Vertex " + vertex + " seen with greedyState = greedy");
            }
            else
            {
                seen[vertex] = GreedyState.nonGreedy;
                log.Debug("Iterator:  Vertex " + vertex + " seen with greedyState = nonGreedy");
            }
        }

        private void AddUnseenChildrenOf(NetworkVertex vertex, bool greedy)
        {
            if (vertex.IsSink) return;
            log.Debug("Iterator: Add unseen children of " + vertex);

            foreach (NetworkEdge edge in graph.EdgesOf(vertex))
            {
                log.Debug("Iterator: Check edge for neighbor in edge " + edge.ToFullInfoString());
                if (!greedy || edge.IsGreedy)
                {
                    NetworkVertex oppositeV = graph.GetOppositeVertex(vertex, edge);//Graphs.getOppositeVertex(graph, edge, vertex);
                    log.Debug("Iterator: Neighbor is " + oppositeV);
                    EncounterVertex(oppositeV, edge);
                }
            }
        }

        private void EncounterStartVertex()
        {
            PutSeenData(startVertex, false);
            stack.Add(startVertex);
            greedyStack.Add(false);
            log.Debug("Iterator: Added to stack " + startVertex + " with greedy set to false");
            startVertexVisited = true;
        }

        private void EncounterVertex(NetworkVertex v, NetworkEdge e)
        {
            if (routeIterator)
            {
                //            if (v == startVertex) return;
                // check the stack
                if (GetCurrentRoute().Contains(v))
                    return;
                // check the seen components 
                //            if (seen.containsKey(v) && (seen.get(v) == greedyState.seen && !v.isSink() 
                //                    || seen.get(v) == greedyState.done || (e.isGreedy() && seen.get(v) == greedyState.nonGreedy)
                //                    || (!e.isGreedy() && seen.get(v) == greedyState.greedy) )) {
                //                log.debug("Do not add vertex " + v  + " to stack");
                //                return;
                //            }
            }
            else
            {
                if (stack.Contains(v)) return;
                if (v.IsSide && seen.ContainsKey(v) && (seen[v] == GreedyState.done || (e.IsGreedy && seen[v] == GreedyState.nonGreedy)
                        || (!e.IsGreedy && seen[v] == GreedyState.greedy)))
                {
                    log.Debug("Leave vertex " + v + " due to greedState rules");
                    return;
                }
            }
            stack.Add(v);
            greedyStack.Add(v.IsSide && !e.IsGreedy);
            log.Debug("Iterator: Added to stack " + v + " with greedy set to " + (v.IsSide && !e.IsGreedy));
        }
    }
}

File: src/main/java/net/sf/rails/algorithms/NetworkIterator.java
package net.sf.rails.algorithms;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.NoSuchElementException;

import net.sf.rails.game.PublicCompany;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.jgrapht.*;
import org.jgrapht.traverse.*;


public class NetworkIterator extends 
    AbstractGraphIterator<NetworkVertex, NetworkEdge> {

    public static enum greedyState {
        seen,
        nonGreedy,
        greedy,
        done
    }
    
    // settings
    private NetworkVertex startVertex;
    private boolean startVertexVisited;
    private boolean routeIterator;

    // internal data
    private List<NetworkVertex> stack = new ArrayList<NetworkVertex>();
    private List<Boolean> greedyStack = new ArrayList<Boolean>();
    private Map<NetworkVertex, greedyState> seen = new HashMap<NetworkVertex, greedyState>();
    

    private final Graph<NetworkVertex, NetworkEdge> graph;

    protected static Logger log =
        LoggerFactory.getLogger(NetworkIterator.class);

    
    public NetworkIterator(Graph<NetworkVertex, NetworkEdge> graph,
            NetworkVertex startVertex) {
        this(graph, startVertex, null);
    }

    /**
     * Returns NetworkIterator for specific company
     */
    public NetworkIterator(Graph<NetworkVertex, NetworkEdge> graph, NetworkVertex startVertex,
            PublicCompany company) {
        super(graph);
        
        if (graph == null)
            throw new IllegalArgumentException("graph must not be null");
        
        if (!graph.containsVertex(startVertex))
            throw new IllegalArgumentException("graph must contain the start vertex");

        this.graph = graph;
        this.startVertex = startVertex;
        this.startVertexVisited = false;
        this.routeIterator = false;
    }
    
    NetworkIterator setRouteIterator(boolean routeIterator) {
        this.routeIterator = routeIterator;
        return this;
    }
    
    /**
     * @return the graph being traversed
     */
    public Graph<NetworkVertex, NetworkEdge> getGraph() {
        return graph;
    }
    
    public Map<NetworkVertex, greedyState> getSeenData() {
        return seen;
    }

    public List<NetworkVertex> getCurrentRoute() {
        // extract all networkvertices just before a null
        List<NetworkVertex> route = new ArrayList<NetworkVertex>();
        NetworkVertex previousVertex = null;
        for (NetworkVertex vertex:stack) {
            if (previousVertex != null && vertex == null) {
                route.add(previousVertex);
            }
            previousVertex = vertex;
        }
        return route;
    }
    
    /**
     * @see java.util.Iterator#hasNext()
     */
    public boolean hasNext() {
        if (!startVertexVisited) {
            encounterStartVertex();
        }
        
        int i = stack.size() - 1;
        while (i >= 0) {
            if (stack.get(i) != null) 
                break;
            else
                i = i - 2;
        }
        return i >=0; 
    }

    /**
     * @see java.util.Iterator#next()
     */
    public NetworkVertex next()
    {
        if (!startVertexVisited) {
            encounterStartVertex();
        }

        if (hasNext()) {
            NetworkVertex nextVertex;
            while (true) {
                 nextVertex = stack.remove(stack.size() - 1);
                if (nextVertex != null)
                    break;
                stack.remove(stack.size() - 1);
            }

            log.debug("Iterator: provides next vertex" + nextVertex);
            boolean nextGreedy = greedyStack.remove(greedyStack.size() - 1);

            putSeenData(nextVertex, nextGreedy);
            stack.add(nextVertex);
            stack.add(null); // add sentinel that we know when we are ready
            
            addUnseenChildrenOf(nextVertex, nextGreedy);

            return nextVertex;
        } else {
            throw new NoSuchElementException();
        }
    }

    
    private void putSeenData(NetworkVertex vertex, boolean greedy) {
        if (!vertex.isSide()) {
            seen.put(vertex, greedyState.seen);
            log.debug("Iterator:  Vertex " + vertex + " seen with greedyState = seen");
            return;
        }
        // side
        if (seen.containsKey(vertex)){
            seen.put(vertex, greedyState.done);
            log.debug("Iterator:  Vertex " + vertex + " seen with greedyState = done");
        } else if (greedy) {
            seen.put(vertex, greedyState.greedy);
            log.debug("Iterator:  Vertex " + vertex + " seen with greedyState = greedy");
        } else {
            seen.put(vertex, greedyState.nonGreedy);
            log.debug("Iterator:  Vertex " + vertex + " seen with greedyState = nonGreedy");
        }
    }
    
    private void addUnseenChildrenOf(NetworkVertex vertex, boolean greedy) {

        if (vertex.isSink()) return;
        log.debug("Iterator: Add unseen children of " + vertex);

        for (NetworkEdge edge : graph.edgesOf(vertex)) {
            log.debug("Iterator: Check edge for neighbor in edge " + edge.toFullInfoString());
            if (!greedy || edge.isGreedy()) {
                NetworkVertex oppositeV = Graphs.getOppositeVertex(graph, edge, vertex);
                log.debug("Iterator: Neighbor is " + oppositeV);
                encounterVertex(oppositeV, edge);
            }
        }
    }
    
    private void encounterStartVertex() {
        putSeenData(startVertex, false);
        stack.add(startVertex);
        greedyStack.add(false);
        log.debug("Iterator: Added to stack " + startVertex +  " with greedy set to false");
        startVertexVisited = true;
    }


    private void encounterVertex(NetworkVertex v, NetworkEdge e){

        if (routeIterator) {
//            if (v == startVertex) return;
            // check the stack
            if (getCurrentRoute().contains(v))
                return;
            // check the seen components 
//            if (seen.containsKey(v) && (seen.get(v) == greedyState.seen && !v.isSink() 
//                    || seen.get(v) == greedyState.done || (e.isGreedy() && seen.get(v) == greedyState.nonGreedy)
//                    || (!e.isGreedy() && seen.get(v) == greedyState.greedy) )) {
//                log.debug("Do not add vertex " + v  + " to stack");
//                return;
//            }
        } else {
            if (stack.contains(v)) return;
            if (v.isSide() && seen.containsKey(v) && (seen.get(v) == greedyState.done || (e.isGreedy() && seen.get(v) == greedyState.nonGreedy)
                    || (!e.isGreedy() && seen.get(v) == greedyState.greedy) )) {
                log.debug("Leave vertex " + v + " due to greedState rules");
                return;
            }
        }
        stack.add(v);
        greedyStack.add(v.isSide() && !e.isGreedy());
        log.debug("Iterator: Added to stack " + v +  " with greedy set to " + (v.isSide() && !e.isGreedy()));
    }

}
